### PR TITLE
[WIP] Improve selectors

### PR DESCRIPTION
**Pull request by kenzierocks**
Selectors need a bit of improvement, as noted by @Minecrell and by me on #355. Still a work in progress, but please comment on the arguments draft as I'm sure it needs improving.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/text/selector/SelectorBuilder.java [hunk 1]
```diff
@@ -36,7 +35,8 @@
     /**
      * Sets the type of this selector.
      *
-     * @param type The type to set
+     * @param type
+     *            The type to set
      * @return This selector builder
      */
     SelectorBuilder selectorType(SelectorType type);
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/text/selector/SelectorBuilder.java]: Could rename this back to `type` now.

File: src/main/java/org/spongepowered/api/text/selector/ArgumentFactory.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    ArgumentBuilder createEmptyBuilder();
+
+    /**
+     * Parses an {@link Argument} from the given argument string.
+     *
+     * @param argument
+     *            The raw argument string
+     * @return A new Argument containing the given argument data
+     */
+    Argument parseRawArgument(String argument);
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/text/selector/ArgumentFactory.java]: What if the argument string is malformed and cannot be parsed? Throw an exception or return `Optional<Argument>`

File: src/main/java/org/spongepowered/api/text/selector/Argument.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+     */
+    Argument invert() throws IllegalStateException;
+
+    /**
+     * Converts this {@linkplain Argument} to a valid argument string.
+     *
+     * @return A valid Argument string that can be parsed with
+     *         {@link Arguments#parseRaw(String)}
+     */
+    String asString();
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/text/selector/Argument.java]: Why not `@Override toString()`?

File: src/main/java/org/spongepowered/api/text/selector/Argument.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ */
+package org.spongepowered.api.text.selector;
+
+/**
+ * Represents a {@linkplain Selector selector} argument. An Argument is a
+ * key-to-value mapping ({@code key=value}) used only for building a selector,
+ * as Map provides a better way of getting this information than iterating over
+ * a list of Arguments.
+ * 
+ * @author Kenzie Togami
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/text/selector/Argument.java]: Accidental `@author` tag

Comment (kenzierocks) [inline src/main/java/org/spongepowered/api/text/selector/Argument.java]: >.< Thanks Eclipse.

File: src/main/java/org/spongepowered/api/text/selector/Argument.java (hunk 3 above)
Comment (kenzierocks) [inline src/main/java/org/spongepowered/api/text/selector/Argument.java]: Same reason that didn't happen in `Selector`, minecrell wants it that way.

**Comment by Minecrell** [comment]
Actually, can you close this pull request and open it again against `feature/text`? As far as I know you can't change that once the pull request was opened so you will probably need to create it again. I've already made some changes there.

File: src/main/java/org/spongepowered/api/text/selector/SelectorBuilder.java (hunk 1 above)
Comment (kenzierocks) [inline src/main/java/org/spongepowered/api/text/selector/SelectorBuilder.java]: Will do.

Comment (Minecrell) [inline src/main/java/org/spongepowered/api/text/selector/SelectorBuilder.java]: This is already in https://github.com/SpongePowered/SpongeAPI/commit/3ecf15584fdf5515f5bd065f708fc7177c6711de

**Comment by kenzierocks** [comment]
mv #404 #405

**Pull request closed without merging**